service/quota.py

```diff
@@ -1,16 +1,106 @@
 from threepio import logger
 
-from core.models import IdentityMembership, Identity, Provider
-
+from core.models import IdentityMembership, Identity
+from core.models.quota import (
+    has_floating_ip_count_quota,
+    has_port_count_quota,
+    has_instance_count_quota,
+    has_cpu_quota,
+    has_mem_quota,
+    has_storage_quota,
+    has_storage_count_quota,
+    has_snapshot_count_quota
+    )
 from service.cache import get_cached_driver
 from service.driver import get_account_driver
 
+
+def check_over_instance_quota(
+        username, identity_uuid,
+        esh_size=None, action=None):
+    """
+    Checks quota based on current limits (and an instance of size, if passed).
+
+    return or raise exc
+    """
+    membership = IdentityMembership.objects.get(
+        identity__uuid=identity_uuid,
+        member__name=username)
+    quota = membership.quota
+    identity = membership.identity
+    driver = get_cached_driver(identity=identity)
+    new_port = new_floating_ip = new_instance = new_cpu = new_ram = 0
+    if esh_size:
+        new_cpu += esh_size.cpu
+        new_ram += esh_size.ram
+    if action in ['launch', 'resume', 'reboot', 'unshelve']:
+        new_floating_ip += 1
+        new_instance += 1
+        new_port += 1
+    # Will throw ValidationError if false.
+    has_cpu_quota(driver, quota, new_cpu)
+    has_mem_quota(driver, quota, new_ram)
+    has_instance_count_quota(driver, quota, new_instance)
+    has_floating_ip_count_quota(driver, quota, new_floating_ip)
+    has_port_count_quota(driver, quota, new_port)
+    return True
+
+
+def check_over_storage_quota(
+        username, identity_uuid,
+        new_snapshot_size=0, new_volume_size=0):
+    """
+    Checks quota based on current limits (and an instance of size, if passed).
+
+    return 5-tuple: ((bool) over_quota,
+                     (str) resource_over_quota,
+                     (int) number_requested,
+                     (int) number_used,
+                     (int) number_allowed)
+    """
+    membership = IdentityMembership.objects.get(identity__uuid=identity_uuid,
+                                                member__name=username)
+    quota = membership.quota
+    identity = membership.identity
+    driver = get_cached_driver(identity=identity)
+    new_disk = new_volume_size or new_snapshot_size
+    new_volume = 1 if new_volume_size > 0 else 0
+    new_snapshot = 1 if new_snapshot_size > 0 else 0
+    # Will throw ValidationError if false.
+    has_storage_quota(driver, quota, new_disk)
+    has_storage_count_quota(driver, quota, new_volume)
+    has_snapshot_count_quota(driver, quota, new_snapshot)
+    return True
+
+
+def set_provider_quota(identity_uuid, limit_dict=None):
+    """
+    """
```

Comment:
Okay - from looking at `_get_hard_limits` the structure of `limit_dict` is just two keys: `"ram", "cpu"`.